synapse/storage/databases/main/event_federation.py

```diff
@@ -151,11 +152,171 @@ async def get_auth_chain_difference(
             The set of the difference in auth chains.
         """
 
-        return await self.db_pool.runInteraction(
-            "get_auth_chain_difference",
-            self._get_auth_chain_difference_txn,
-            state_sets,
-        )
+        # Check if we have indexed the room so we can use the chain cover
+        # algorithm.
+        room = await self.get_room(room_id)
+        if room["has_auth_chain_index"]:
+            return await self.db_pool.runInteraction(
+                "get_auth_chain_difference_chains",
+                self._get_auth_chain_difference_using_chains_txn,
```

Comment:
Maybe avoiding using `chains` again would make this clearer? Perhaps something like "get_auth_chain_difference_via_cover_index"?